nupic/research/frameworks/greedy_infomax/utils/model_utils.py

```diff
@@ -0,0 +1,335 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2021, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+#
+# This work was based on the original Greedy InfoMax codebase from Sindy Lowe:
+# https://github.com/loeweX/Greedy_InfoMax
+# The Greedy InfoMax paper can be found here:
+# https://arxiv.org/abs/1905.11786
+# ----------------------------------------------------------------------
+
+import sys
+import time
+
+import torch
+import torch.nn.functional as F
+from tqdm import tqdm
+
+
+def gen_orthgonal(dim):
+    a = torch.zeros((dim, dim)).normal_(0, 1)
+    q, r = torch.qr(a)
+    d = torch.diag(r, 0).sign()
+    diag_size = d.size(0)
+    d_exp = d.view(1, diag_size).expand(diag_size, diag_size)
+    q.mul_(d_exp)
+    return q
+
+
+def make_delta_orthogonal(weights, gain):
+    rows = weights.size(0)
+    cols = weights.size(1)
+    if rows > cols:
+        print("In_filters should not be greater than out_filters.")
+    weights.data.fill_(0)
+    dim = max(rows, cols)
+    q = gen_orthgonal(dim)
+    mid1 = weights.size(2) // 2
+    mid2 = weights.size(3) // 2
+    with torch.no_grad():
+        weights[:, :, mid1, mid2] = q[: weights.size(0), : weights.size(1)]
+        weights.mul_(gain)
+
+
+def patchify_inputs(x, patch_size, overlap):
+    x = (
+        x.unfold(2, patch_size, patch_size // overlap)
+        .unfold(3, patch_size, patch_size // overlap)
+        .permute(0, 2, 3, 1, 4, 5)  # b, p_x, p_y, c, x, y
+    )
+    n_patches_x = x.shape[1]
+    n_patches_y = x.shape[2]
+    x = x.reshape(
+        x.shape[0] * x.shape[1] * x.shape[2], x.shape[3], x.shape[4], x.shape[5]
+    )
+    return x, n_patches_x, n_patches_y
+
+
+def train_model_gim(
+    model,
+    loader,
+    optimizer,
```

Comment:
Style comment: to be consistent with other train/eval functions, consider changing the name here to `train_gim_model` (and likewise for the eval function below). This would be consistent with how we defined [train & eval functions for dendrites](https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/dendrites/model_utils.py#L33-L50).

Reply:
Done